From 974c43a620c524e038972acfdc73b866976ffbba Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 8 Sep 2014 13:09:02 -0400
Subject: [tools] perf: Add global count of opened dso objects

Message-id: <1410181752-25631-298-git-send-email-jolsa@redhat.com>
Patchwork-id: 91864
O-Subject: [PATCH RHEL7.1 BZ1134356 297/307] perf tools: Add global count of opened dso objects
Bugzilla: 1134356
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1134356
https://bugzilla.redhat.com/show_bug.cgi?id=1134356

upstream
========
commit bda6ee4a94d1e1be0c1428d37bc0d3da2e5793ad
Author: Jiri Olsa <jolsa@kernel.org>
Date: Wed Apr 30 15:25:10 2014 +0200

description
===========
Adding global count of opened dso objects so we could
properly limit the number of opened dso data file
descriptors.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/dso.c b/tools/perf/util/dso.c
index 5d7c7bc..76e5c13 100644
--- a/tools/perf/util/dso.c
+++ b/tools/perf/util/dso.c
@@ -1,3 +1,4 @@
+#include <asm/bug.h>
 #include "symbol.h"
 #include "dso.h"
 #include "machine.h"
@@ -137,18 +138,23 @@ int dso__read_binary_type_filename(const struct dso *dso,
 }
 
 /*
- * Global list of open DSOs.
+ * Global list of open DSOs and the counter.
  */
 static LIST_HEAD(dso__data_open);
+static long dso__data_open_cnt;
 
 static void dso__list_add(struct dso *dso)
 {
  list_add_tail(&dso->data.open_entry, &dso__data_open);
+ dso__data_open_cnt++;
 }
 
 static void dso__list_del(struct dso *dso)
 {
  list_del(&dso->data.open_entry);
+ WARN_ONCE(dso__data_open_cnt <= 0,
+    "DSO data fd counter out of bounds.");
+ dso__data_open_cnt--;
 }
 
 static int __open_dso(struct dso *dso, struct machine *machine)
-- 
1.7.1